Test2

tingstad opened the pull request:
Rewrite shell test to avoid `sleep` and polling. Ref https://github.com/teodorlu/hotload/pull/5

The output from hotload.py is piped to a subshell command group 
containing a sequence of `expect` function calls.
These functions assert that the expected strings are observered, in order. (Ignoring other output interspersed.)

After running the test, we want to tidy up and end the python process. To get its PID we need to use a named pipe, `"$pipe"`.

To avoid infinite "hang", a timeout is set by running `sleep` in parallell with the tests. Again a named pipe is needed, in this case to wait for the first of the jobs (tests, sleep) to finish. The main process reads from FIFO pipe `$done_chan`, blocking until a message is received.

The test runs pretty fast: https://github.com/teodorlu/hotload/runs/7297455043

```
2022-07-12T07:58:27.8053635Z ##[group]Run sh -x .github/workflows/test.sh
...
2022-07-12T07:58:28.2685395Z Tests OK!
```
=> 0.463 seconds. About 0.09s on my laptop. Remember, this includes starting python and several other processes (sed etc.).

tingstad committed:
Add shell test

tingstad committed:
debug

tingstad committed:
Test: shell w/o sleep/poll

Rewrite shell test to avoid `sleep` and polling.

The output from hotload.py is piped to a subshell command group
containing a sequence of `expect` function calls.
These functions assert that the expected strings are observered, in order.

After running the test, we want to tidy up and end the python process. To get its PID we need to use a named pipe, `"$pipe"`.

To avoid infinite "hang", a timeout is set by running `sleep` in parallell with the tests. Again a named pipe is needed, in this case to wait for the first of the jobs (tests, sleep) to finish. The main process reads from FIFO pipe `$done_chan`, blocking until a message is received.

tingstad committed:
Test: restructure code

Move util functions furhter down.

teodorlu commented:
Denne går gjennom med returkode 0 hos meg!

<details><summary><code>$ sh -x .github/workflows/test.sh</code></summary>
<pre><code>
+ set -e
+ main
+ done_chan=chan44542
+ mkfifo chan44542
+ pid1=44544
+ tests chan44542
+ pid2=44545
+ done_chan=chan44542
+ read msg
+ trap end EXIT
+ cat
+ sleep 4
+ touch -t 200711121015 lib.py
+ pipe=hotload44542
+ mkfifo hotload44542
+ pid=44551
+ echo lib.py
+ python -u hotload.py lib.py
+ expect 'Running hotload'
+ IFS=
+ read -r line
+ case "$line" in
+ echo 'Asserted: Running hotload'
Asserted: Running hotload
+ return 0
+ expect 'Result: 25'
+ IFS=
+ read -r line
+ case "$line" in
+ IFS=
+ read -r line
+ case "$line" in
+ IFS=
+ read -r line
+ case "$line" in
+ IFS=
+ read -r line
+ case "$line" in
+ IFS=
+ read -r line
+ case "$line" in
+ echo 'Asserted: Result: 25'
Asserted: Result: 25
+ return 0
+ expect 'Successfully reloaded lib'
+ IFS=
+ read -r line
+ case "$line" in
+ IFS=
+ read -r line
+ case "$line" in
+ echo 'Asserted: Successfully reloaded lib'
Asserted: Successfully reloaded lib
+ return 0
+ sed -i.bak 's/x = ./x = 4/' lib.py
+ expect 'Result: 32'
+ IFS=
+ read -r line
+ case "$line" in
+ echo 'Asserted: Result: 32'
Asserted: Result: 32
+ return 0
+ expect 'Successfully reloaded lib'
+ IFS=
+ read -r line
+ case "$line" in
+ echo 'Asserted: Successfully reloaded lib'
Asserted: Successfully reloaded lib
+ return 0
+ echo OK
++ end
++ '[' -z 44551 ']'
++ nofail silent kill 44551
++ silent kill 44551
++ kill 44551
+ silent nofail kill 44544 44545
++ '[' -e hotload44542 ']'
+ nofail kill 44544 44545
++ rm hotload44542
+ case "$msg" in
+ echo 'Tests OK!'
Tests OK!
+ exit

</code></pre>
</details>


```
$ time sh -x .github/workflows/test.sh 1>/dev/null 2>&1
sh -x .github/workflows/test.sh > /dev/null 2>&1  0,01s user 0,00s system 3% cpu 0,174 total
```

teodorlu commented:
TMPDIR er ikke satt hos meg, så jeg får et par chan* og hotload*-filer i toppmappa.

Merk - helt OK for meg å merge som den er. Derfor approve. Bye bedre å ha CI enn å ikke ha CI.

---

Er dette mat for `mktemp -d`?

Feks som

```
tempdir=$(mktmp -d)
# Så bruke denne til `done_chan` og `pipe`?
```

```
teodorlu@teod-t490s ~/dev/teodorlu/hotload (git)-[remotes/tingstad/test2] % ls -l
total 44
-rw-r--r-- 1 teodorlu teodorlu 9323 2022-07-03 21:54 README.md
prw-r--r-- 1 teodorlu teodorlu    0 2022-07-12 13:11 chan44100
prw-r--r-- 1 teodorlu teodorlu    0 2022-07-12 13:12 chan44238
prw-r--r-- 1 teodorlu teodorlu    0 2022-07-12 13:12 chan44405
prw-r--r-- 1 teodorlu teodorlu    0 2022-07-12 13:12 chan44542
prw-r--r-- 1 teodorlu teodorlu    0 2022-07-12 13:14 chan44755
prw-r--r-- 1 teodorlu teodorlu    0 2022-07-12 13:14 chan44888
drwxr-xr-x 2 teodorlu teodorlu 4096 2022-01-06 11:44 dev
drwxr-xr-x 3 teodorlu teodorlu 4096 2022-05-06 13:29 docs
-rw-r--r-- 1 teodorlu teodorlu  218 2022-05-06 14:50 example_script.py
-rwxr-xr-x 1 teodorlu teodorlu 7935 2022-07-12 13:10 hotload.py
prw-r--r-- 1 teodorlu teodorlu    0 2022-07-12 13:11 hotload44100
prw-r--r-- 1 teodorlu teodorlu    0 2022-07-12 13:12 hotload44238
prw-r--r-- 1 teodorlu teodorlu    0 2022-07-12 13:12 hotload44405
prw-r--r-- 1 teodorlu teodorlu    0 2022-07-12 13:12 hotload44542
prw-r--r-- 1 teodorlu teodorlu    0 2022-07-12 13:14 hotload44755
prw-r--r-- 1 teodorlu teodorlu    0 2022-07-12 13:14 hotload44888
-rw-r--r-- 1 teodorlu teodorlu  283 2022-01-06 11:44 install.sh
-rw-r--r-- 1 teodorlu teodorlu   50 2022-07-12 13:14 lib.py
drwxr-xr-x 2 teodorlu teodorlu 4096 2022-07-12 13:14 __pycache__
```

tingstad commented:
> TMPDIR er ikke satt hos meg, så jeg får et par chan* og hotload*-filer i toppmappa.
> 
> Merk - helt OK for meg å merge som den er. Derfor approve. Bye bedre å ha CI enn å ikke ha CI.
> 
> Er dette mat for `mktemp -d`?

1. Jeg oppdaget at ikke alle prosesser ble ryddet skikkelig ved (hypotetiske) feil. Ingen trap på toppnivå.
2. Pen fiks av 1 virket dårlig med GitHub som kjører script i samme process group som runner.
3. La inn sletting av chan-filene i sammen med prosess-rydde-koden.

Alt dette er fikset i https://github.com/teodorlu/hotload/pull/12
Som antydet ble den litt ... interessant, så kanskje du vil teste på nytt.

`mktemp` er ikke standardisert, så ville unngå den, men endte uansett opp med Bash-krav (se PR 12). Men nå sletter jeg filene, så jeg tror det vil være bra.

tingstad committed:
Test3 - apocalypse (#12)

"Test2", https://github.com/teodorlu/hotload/pull/11, works pretty well, but does not clean up all processes on errors.

### Traps galore — Signal madness 

Make sure all child processes are killed when script ends, both when successful and otherwise.

Put cleanup code in a single top-level function, `end()`. So far so good.

Send kill signal to all children by getting pgid (process group id) of all jobs.

The script should quit with the correct exit code, so we have to set `trap 'exit $?' TERM`.

We cannot send SIGINT to the asynchronous commands (it's ignored when job control is disabled),
so TERM is sent, and `trap 'trap - TERM; kill -s INT $$' TERM` is set for the parent process.

### PID armageddon! 

Kill each PID decentant.

GitHub runs the script in the same process group as the runner, so killing the process group is problematic.

This commit kills every descendant PID instead.

Not sure if the order matters, so to be on the safe side the youngest PIDs are printed first.

I.e. we call `kill GRANDCHILD1 GRANDCHILD2 CHILD1 CHILD2 PARENT`

The function prints lines of:

    ...
    GREAT ** N  GRANDCHILDREN  (N >= 2)
    GREAT GRANDCHILDREN
    GRANDCHILDREN
    CHILDREN
    PARENT (`root`)

#### kill $pid and cleanup fifos 820c45c

On successful runs, $pid did not terminate, getting parent PID 1.

Add EXIT trap to terminate $pid, and also delete FIFOs.

The EXIT trap is called on all exits in Bash, including SIGTERM.

Other shells runs EXIT trap only in the absence of signals.

Keeping this generic is too much work, the shebang reflects Bash requirement.
